### ISSUE TITLE
OkHttp: Exoplayer bitrate estimate is not accurate

### ISSUE THREAD

**Issue by sridhard** [google/ExoPlayer#5038]
### Issue description
Exoplayer bitrate estimate is not accurate

### Reproduction steps
I have implemented the interface  onBandwidthSample of BandwidthMeter.EventListener and passed to DefaultBandwidthMeter as below:

```
private void buildBandwidthMeter() {
        BANDWIDTH_METER = new DefaultBandwidthMeter(new Handler(), new BandwidthMeter.EventListener() {
            @Override
            public void onBandwidthSample(int elapsedMs, long bytes, long bitrate) {
                System.out.println("TEST::::onBandwidthSample!!!!!!! " +elapsedMs + " " + bytes + " " + bitrate + " " + bw);
            }
        });
    }
```

Network speed is around 2-3Mbps. But the bitrate reported by onBandwidthSample is 20-30Mbps. This is because the elapsedMs is very low in the range of 30-100 ms.

I have installed a local http server to verify the response time of HTTP requests. The actual response time is around 600-1000ms, But elapsedMs is onBandwidthSample is very low. This make the bitrate very high

### Link to test content
I am using local HTTP server so the link cannot be shared. 

### Version of ExoPlayer being used
exoplayer-2.9

### Device(s) and version(s) of Android being used
Xiaomi-mi-max2. android version: 7.1.1

### A full bug report captured from the device
There is no error in the logcat. Also no information is printed related to network transfer rate

**Comment by tonihei** [comment]
For reference - that's a copy of this [post](https://github.com/google/ExoPlayer/issues/5021#issuecomment-434690544).

**Comment by tonihei** [comment]
As said on the other thread, there may be inaccuracies in the estimates during start-up. We need to investigate a bit further to know what needs to be done here. This may also be specific to OkHttpDataSource. I'll mark this issue as an enhancement until this is done. 

As we may not get around to this soon, you can help us if you wish by providing measurements of the current time in onTransferInitializing, onTransferStart, onBytesTransferred and onTransferEnded. And also the number of bytes reported in onBytesTransferred. This should give a better idea of the behaviour.

**Comment by sridhard** [comment]
Hi,

I checked the code. In OkHttpDataSource.java, the http request is made synchronously. The transferStarted() is triggered once execute() is success. This means the the transferStarted() notification time is not correct and as a result the time difference between transferStarted() and transferEnded() is very low. I have added some debug logs when transferStarted() is called and when transferEnded() is called. Also I added debug logs in the onBandwidthSample() of the listner. Below are the logs:

```
11-01 14:07:58.186 18965 19566 I System.out: Debug::: okHttpdataSource.transferStarted() called at 1541061478186
11-01 14:07:58.269 18965 19566 I System.out: Debug::: okHttpdataSource.transferEnded() called at 1541061478269
11-01 14:07:58.270 18965 18965 I System.out: Debug::::In function onBandwidthSample,  elapsedMs = 83  bytes = 146832  bitrate = 13746604

11-01 14:07:58.937 18965 19566 I System.out: Debug::: okHttpdataSource.transferStarted() called at 1541061478937
11-01 14:07:59.055 18965 19566 I System.out: Debug::: okHttpdataSource.transferEnded() called at 1541061479055
11-01 14:07:59.056 18965 18965 I System.out: Debug::::In function onBandwidthSample,  elapsedMs = 118  bytes = 142144  bitrate = 13746604

11-01 14:08:00.024 18965 19566 I System.out: Debug::: okHttpdataSource.transferStarted() called at 1541061480024
11-01 14:08:00.145 18965 19566 I System.out: Debug::: okHttpdataSource.transferEnded() called at 1541061480145
11-01 14:08:00.146 18965 18965 I System.out: Debug::::In function onBandwidthSample,  elapsedMs = 120  bytes = 270544  bitrate = 14124137
```
From the above logs we can see that the difference between transferStarted time and transferEnded time(In millisec) is very less. As a result the bitrate estimate is very high


I move the transferStarted() function before calling execute() function in OkHttpdataSource.java as shown below:

  ```
@Override
  public long open(DataSpec dataSpec) throws HttpDataSourceException {
    

    this.dataSpec = dataSpec;
    this.bytesRead = 0;
    this.bytesSkipped = 0;
    transferInitializing(dataSpec);

    Request request = makeRequest(dataSpec);
    Response response;
    ResponseBody responseBody;

    **System.out.println("Debug::: okHttpdataSource.transferStarted() called at " + System.currentTimeMillis());
    transferStarted(dataSpec);**
    
    try {
      this.response = callFactory.newCall(request).execute();
      response = this.response;
      responseBody = Assertions.checkNotNull(response.body());
      responseByteStream = responseBody.byteStream();
    } catch (IOException e) {
      **System.out.println("TEST::: transferEnded " + System.currentTimeMillis());
      transferEnded();**
      throw new HttpDataSourceException(
          "Unable to connect to " + dataSpec.uri, e, dataSpec, HttpDataSourceException.TYPE_OPEN);
    }

    int responseCode = response.code();

    // Check for a valid response code.
    if (!response.isSuccessful()) {
      **System.out.println("TEST::: transferEnded " + System.currentTimeMillis());
      transferEnded();**

      Map<String, List<String>> headers = response.headers().toMultimap();
      closeConnectionQuietly();
      InvalidResponseCodeException exception = new InvalidResponseCodeException(
          responseCode, headers, dataSpec);
      if (responseCode == 416) {
        exception.initCause(new DataSourceException(DataSourceException.POSITION_OUT_OF_RANGE));
      }
      throw exception;
    }

    // Check for a valid content type.
    MediaType mediaType = responseBody.contentType();
    String contentType = mediaType != null ? mediaType.toString() : "";
    if (contentTypePredicate != null && !contentTypePredicate.evaluate(contentType)) {
      **System.out.println("TEST::: transferEnded " + System.currentTimeMillis());
      transferEnded();**

      closeConnectionQuietly();
      throw new InvalidContentTypeException(contentType, dataSpec);
    }

    // If we requested a range starting from a non-zero position and received a 200 rather than a
    // 206, then the server does not support partial requests. We'll need to manually skip to the
    // requested position.
    bytesToSkip = responseCode == 200 && dataSpec.position != 0 ? dataSpec.position : 0;

    // Determine the length of the data to be read, after skipping.
    if (dataSpec.length != C.LENGTH_UNSET) {
      bytesToRead = dataSpec.length;
    } else {
      long contentLength = responseBody.contentLength();
      bytesToRead = contentLength != -1 ? (contentLength - bytesToSkip) : C.LENGTH_UNSET;
    }

    opened = true;

    return bytesToRead;
  }
```

After the code changes, the logs are below:

```
11-01 14:19:32.906 21663 21958 I System.out: Debug::: okHttpdataSource.transferStarted() called at 1541062172906
11-01 14:19:33.910 21663 21958 I System.out: Debug::: okHttpdataSource.transferEnded() called at 1541062173910
11-01 14:19:33.911 21663 21663 I System.out: Debug::::In function onBandwidthSample,  elapsedMs = 1004  bytes = 143824  bitrate = 1253481

11-01 14:19:33.913 21663 21958 I System.out: Debug::: okHttpdataSource.transferStarted() called at 1541062173913
11-01 14:19:34.767 21663 21958 I System.out: Debug::: okHttpdataSource.transferEnded() called at 1541062174767
11-01 14:19:34.769 21663 21663 I System.out: Debug::::In function onBandwidthSample,  elapsedMs = 854  bytes = 138944  bitrate = 1253481

11-01 14:19:34.770 21663 21958 I System.out: Debug::: okHttpdataSource.transferStarted() called at 1541062174769
11-01 14:19:35.820 21663 21958 I System.out: Debug::: okHttpdataSource.transferEnded() called at 1541062175820
11-01 14:19:35.821 21663 21663 I System.out: Debug::::In function onBandwidthSample,  elapsedMs = 1051  bytes = 145520  bitrate = 1253481
```

From the logs we can see that the elapsedMs is more now and as a result the bitrate estimate is accurate.

I have verified both the test cases in same network and within 10min. Also I have compared the bandwidth with other bandwidth detection apps.

Before code change the bandwidth reported in almost 5-10 times of the actual bandwidth. After code change the bandwidth is more accurate and it matches the actual network conditions.

**Comment by sridhard** [comment]
@tonihei please review the code changes of the above function. I have tested with local HTTP server and also with the CDN. The code works fine.

If the code is OK I will raise a pull request

**Comment by sridhard** [comment]
@tonihei in failure case we need a better way to handle transferEnded(). In failure case we need to report failure status with transferEnded() so that the caller/listener dont consider this request as success and calculate bandwidth

**Comment by tonihei** [comment]
Thanks for the analysis. We still need to think about how to solve this properly. While I can see how the execute() call already seems to take the majority of the time, we should still distinguish between the time to open the connection and the time to load data (if possible). Also, the transferEnded call must happen in close(), that's just how the interface works at the moment. 

How did you configure the OkHttpClient passed to the data source? I wonder if there is any caching involved?

**Comment by sridhard** [comment]
I am using react-native-video plugin to support video in react-native with exoplayer. Below is the function which build the OkHttpClient

   ```
 private static HttpDataSource.Factory buildHttpDataSourceFactory(ReactContext context, DefaultBandwidthMeter bandwidthMeter, Map<String, String> requestHeaders) {
        OkHttpClient client = OkHttpClientProvider.getOkHttpClient();
        CookieJarContainer container = (CookieJarContainer) client.cookieJar();
        ForwardingCookieHandler handler = new ForwardingCookieHandler(context);
        container.setCookieJar(new JavaNetCookieJar(handler));
        OkHttpDataSourceFactory okHttpDataSourceFactory = new OkHttpDataSourceFactory(client, getUserAgent(context), bandwidthMeter);

        if (requestHeaders != null)
            okHttpDataSourceFactory.getDefaultRequestProperties().set(requestHeaders);

        return okHttpDataSourceFactory;
    }
```

As far as I know there is no caching. This I am confident because I uninstalled and reinstalled the app during testing. Also I have verified with different HLS urls.

**Comment by sridhard** [comment]
As per your comment "Also, the transferEnded call must happen in close(), that's just how the interface works at the moment"

Since the http is fetched in synchronous method, we need to call transferStarted() before execute(). After calling execute the request can fail or pass. If the request fails currently close() is not called, but we need to indicate transferEnded().

Any better way to solve this?

**Comment by tonihei** [comment]
As far as I understand .execute() returns as soon as the response is ready (=connection is open). That doesn't mean that the entire stream has been loaded into memory. It still needs to be read using the byte stream obtained from the response. 

Not sure how to improve the current situation yet. I think it needs some further analysis and testing to see what's happening.

**Comment by sridhard** [comment]
In OkHttp there are 2 ways to download a file. Synchronous and asynchronous. More info can be found at https://www.journaldev.com/13629/okhttp-android-example-tutorial

In synchronous method, once the execute() function returns the response will have some data. This means that it has downloaded the data from server. May be execute() will return after receiving the first chunk of data(I am not sure).

exoplayer uses synchronous method. That is the reason we get low time when we call transferStarted() after execute is returned.

**Comment by tonihei** [comment]
If you have the time to investigate, you could try to use the asynchronous callback to see if the timing is better? Otherwise this needs to wait a while until we get around to investigate this.

**Comment by sridhard** [comment]
Ok..I will try it next week.

**Comment by anhlevfx** [comment]
Hello, I'm facing the problem too, but only with HLS live stream. With HLS VOD, get estimate bitrate from exoplayer instance working well.

**Comment by tonihei** [comment]
I justed tried to investigate the issue a bit and wasn't able to reproduce the effect in any way in our demo app. I tried to used OkHttpDataSource with and without OkHttp cache, on a slow network (400-600kbps) and a super-fast network (>100Mbps). Also tried to change to code to use the synchronous or asynchrous OkHttp method calls as suggested in this thread and also tested various stream types (DASH, HLS, both encrypted and clear).

In all cases, the measured bandwidth (from the time the connection was fully open to closing the connection) was within expectations. Actually, including the time to open the connection as proposed in https://github.com/sridhard/ExoPlayer/commit/b4a5b9f3256350bd01598973c28a8a4e470924ed would make the estimations worse. For example, on the slow network many observations were in the correct range of 400-600kbps, whereas including the connection opening time reduced the estimates to around 200-300kbps.

Can anyone provide reproduction steps for the ExoPlayer demo app? It may be that some configuration of the react-native-plugin version is causing these problems.

Note: Including the time to open the connection in the bandwidth estimate or even providing a separate estimate for that might be helpful to improve track selection in the future. However, that goes beyond the problem described here.
